src/main/java/appeng/container/implementations/ContainerFluidTerminal.java

```diff
@@ -0,0 +1,479 @@
+				return;
+			}
+
+			if( isBucket )
+			{
+				// We need to handle buckets separately
+				ItemStack filledBucket = FluidUtil.getFilledBucket( pulled.getFluidStack() );
+				player.inventory.setItemStack( filledBucket );
+			}
+			else
+			{
+				fh.fill( pulled.getFluidStack(), true );
+			}
+			this.updateHeld( player );
+		}
+		else if( action == InventoryAction.EMPTY_ITEM )
+		{
+			// Empty held item
+			AELog.info( "Emptying %s", held.getDisplayName() );
+
+			// See how much we can drain from the item
+			FluidStack extract = fh.drain( Integer.MAX_VALUE, false );
+			if( extract == null || extract.amount < 1 )
+			{
+				return;
+			}
+
+			// Check if we can push into the system
+			IAEFluidStack canPush = this.monitor.injectItems( AEFluidStack.fromFluidStack( extract ), Actionable.SIMULATE, this.getActionSource() );
+			if( isBucket && canPush != null && canPush.getStackSize() > 0 )
+			{
+				// We can't push enough for the bucket
+				return;
+			}
+
+			IAEFluidStack inserted = Platform.poweredInsert( this.getPowerSource(), this.monitor, AEFluidStack.fromFluidStack( extract ), this.getActionSource() );
+			if( inserted != null && inserted.getStackSize() > 0 )
+			{
+				// Only try to extract the amount we DID insert
+				extract.amount -= Math.toIntExact( inserted.getStackSize() );
+			}
+
+			if( isBucket )
+			{
+				// Remove bucket and replace with EmptyBucket
+				player.inventory.setItemStack( new ItemStack( Items.BUCKET, 1 ) );
+			}
+			else
+			{
+				// Actually drain
+				fh.drain( extract, true );
+			}
+			this.updateHeld( player );
+		}
+	}
+
+	public void setTargetStack( final IAEFluidStack stack )
+	{
+		if( Platform.isClient() )
+		{
+			if( stack == null && this.clientRequestedTargetFluid == null )
+			{
+				return;
+			}
+			if( stack != null && this.clientRequestedTargetFluid != null && stack.getFluidStack().isFluidEqual( this.clientRequestedTargetFluid.getFluidStack() ) )
+			{
+				return;
+			}
+			NetworkHandler.instance().sendToServer( new PacketTargetFluidStack( (AEFluidStack) stack ) );
+		}
+
+		this.clientRequestedTargetFluid = stack == null ? null : stack.copy();
+	}
+
+	@Override
+	public void updateSetting( IConfigManager manager, Enum settingName, Enum newValue )
+	{
+		if( this.getGui() != null )
+		{
+			this.getGui().updateSetting( manager, settingName, newValue );
+		}
+	}
+
+	@Override
+	public void detectAndSendChanges()
+	{
+		if( Platform.isServer() )
+		{
+			if( this.monitor != this.terminal.getInventory( AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class ) ) )
+			{
+				this.setValidContainer( false );
+			}
+
+			for( final Settings set : this.serverCM.getSettings() )
+			{
+				final Enum<?> sideLocal = this.serverCM.getSetting( set );
+				final Enum<?> sideRemote = this.clientCM.getSetting( set );
+
+				if( sideLocal != sideRemote )
+				{
+					this.clientCM.putSetting( set, sideLocal );
+					for( final IContainerListener crafter : this.listeners )

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)